File: array/SlidingWindow/variable_size/cpp/SlidingWindow.cpp
#include "SlidingWindow.hpp"

using namespace std;

namespace {
bool CheckedAdd(long long a, long long b, long long& result) {
    if ((b > 0 && a > 9223372036854775807LL - b) ||
        (b < 0 && a < (-9223372036854775807LL - 1LL) - b)) {
        return false;
    }
    result = a + b;
    return true;
}

bool CheckedSub(long long a, long long b, long long& result) {
    if ((b < 0 && a > 9223372036854775807LL + b) ||
        (b > 0 && a < (-9223372036854775807LL - 1LL) + b)) {
        return false;
    }
    result = a - b;
    return true;
}
}  // namespace

bool MinWindowAtLeastTarget(const vector<long long>& values,
                            long long target,
                            int& bestLength,
                            int& bestLeft,
                            int& bestRight) {
    long long windowSum = 0;
    int left = 0;
    bestLength = -1;
    bestLeft = -1;
    bestRight = -1;

    for (int right = 0; right < static_cast<int>(values.size()); ++right) {
        if (!CheckedAdd(windowSum, values[right], windowSum)) {
            return false;
        }

        while (windowSum >= target) {
            const int currentLength = right - left + 1;
            if (bestLength == -1 || currentLength < bestLength) {
                bestLength = currentLength;
                bestLeft = left;
                bestRight = right;
            }

            if (!CheckedSub(windowSum, values[left], windowSum)) {
                return false;
            }
            ++left;
        }
    }

    return true;
}

File: array/SlidingWindow/variable_size/rust/src/sliding_window.rs
pub fn min_window_at_least_target(values: &[i64], target: i64) -> Option<(i32, i32, i32)> {
    let mut window_sum = 0_i64;
    let mut left = 0_usize;
    let mut best_length = -1_i32;
    let mut best_left = -1_i32;
    let mut best_right = -1_i32;

    for (right, &value) in values.iter().enumerate() {
        window_sum = window_sum.checked_add(value)?;

        while window_sum >= target {
            let current_length = (right - left + 1) as i32;
            if best_length == -1 || current_length < best_length {
                best_length = current_length;
                best_left = left as i32;
                best_right = right as i32;
            }
            window_sum = window_sum.checked_sub(values[left])?;
            left += 1;
        }
    }

    Some((best_length, best_left, best_right))
}
